### ISSUE TITLE
MSVC: Destination must be a directory at .\util\copy.pl line 53.

### ISSUE THREAD

**Issue by misery** [openssl/openssl#11623]
Tried openssl-3.0.0-alpha1 on all platforms of our CI server. Works good but MSVC build cannot install anymore. Same configuration works with 1.1.1g.

```
Copying: ./include/openssl/x509v3err.h to C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/include/openssl/x509v3err.h
created directory `C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/lib'
Copying: libcrypto.lib to C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/lib/libcrypto.lib
Copying: libssl.lib to C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/lib/libssl.lib
	"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\Hostx86\x86\nmake.exe" /                   depend && "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\Hostx86\x86\nmake.exe" /                   _build_modules

Microsoft (R) Program Maintenance Utility, Version 14.16.27030.1
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.


Microsoft (R) Program Maintenance Utility, Version 14.16.27030.1
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

*** Installing engines
created directory `C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/lib/engines-3'
*** Installing modules
created directory `C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist/lib/ossl-modules'
Destination must be a directory at .\util\copy.pl line 53.
Die Syntax f�r den Dateinamen, Verzeichnisnamen oder die Datentr„gerbezeichnung ist falsch.
NMAKE : fatal error U1077: "if": Rückgabe-Code "0x1"
Stop.
NMAKE : fatal error U1077: "echo": Rückgabe-Code "0x2"
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\nmake.exe"": Rückgabe-Code "0x2"
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\nmake.exe"": Rückgabe-Code "0x2"
Stop.
NMAKE : fatal error U1077: ""C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\nmake.exe"": Rückgabe-Code "0x2"
Stop.
```

**Configuration**
```
    C:\Perl64\bin\perl.exe Configure --prefix=C:/J/workspace/default_Review_Libs_Win32_MSVC/build/dist no-ui-console no-filenames no-camellia no-bf no-aria no-seed no-poly1305 no-srp no-gost no-idea no-mdc2 no-rc2 no-rc4 no-rc5 no-srtp no-sm2 no-sm3 no-sm4 no-ct no-dgram no-cast no-chacha no-blake2 no-rmd160 no-scrypt no-siphash no-whirlpool no-md4 no-des no-tls1 no-tls1-method no-tls1_1 no-tls1_1-method no-tls1_3 no-ssl3 no-ssl3-method no-dtls no-dtls1-method no-dtls1_2-method no-engine no-async no-dso no-comp no-ts no-makedepend no-tests no-legacy shared no-asm VC-WIN32 -Os

Perl information:

    C:\Perl64\bin\perl.exe
    5.26.3 for MSWin32-x64-multi-thread

Enabled features:

    autoalginit
    autoerrinit
    autoload-config
    cmac
    cmp
    cms
    deprecated
    dh
    dsa
    ec
    ec2m
    ecdh
    ecdsa
    err
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    pic
    posix-io
    psk
    rdrand
    rfc3779
    secure-memory
    shared
    siv
    sock
    sse2
    ssl
    static-engine
    stdio
    threads
    tls
    uplink
    tls1_2
    tls1_2-method

Disabled features:

    afalgeng            [cascade]                OPENSSL_NO_AFALGENG
    aria                [option]                 OPENSSL_NO_ARIA (skip crypto\aria)
    asan                [default]                OPENSSL_NO_ASAN
    asm                 [option]                 OPENSSL_NO_ASM
    async               [option]                 OPENSSL_NO_ASYNC
    bf                  [option]                 OPENSSL_NO_BF (skip crypto\bf)
    blake2              [option]                 OPENSSL_NO_BLAKE2
    buildtest-c++       [default]                
    camellia            [option]                 OPENSSL_NO_CAMELLIA (skip crypto\camellia)
    capieng             [cascade]                OPENSSL_NO_CAPIENG
    cast                [option]                 OPENSSL_NO_CAST (skip crypto\cast)
    chacha              [option]                 OPENSSL_NO_CHACHA (skip crypto\chacha)
    comp                [option]                 OPENSSL_NO_COMP (skip crypto\comp)
    crypto-mdebug       [default]                OPENSSL_NO_CRYPTO_MDEBUG
    ct                  [option]                 OPENSSL_NO_CT (skip crypto\ct)
    des                 [option]                 OPENSSL_NO_DES (skip crypto\des)
    devcryptoeng        [default]                OPENSSL_NO_DEVCRYPTOENG
    dgram               [option]                 OPENSSL_NO_DGRAM
    dso                 [option]                 OPENSSL_NO_DSO
    dtls                [option(dtls)]           OPENSSL_NO_DTLS
    dynamic-engine      [cascade]                
    ec_nistp_64_gcc_128 [default]                OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]                OPENSSL_NO_EGD
    engine              [option]                 OPENSSL_NO_ENGINE (skip engines, crypto\engine)
    external-tests      [default]                OPENSSL_NO_EXTERNAL_TESTS
    filenames           [option]                 OPENSSL_NO_FILENAMES
    fips                [cascade]                
    fuzz-libfuzzer      [default]                OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl            [default]                OPENSSL_NO_FUZZ_AFL
    gost                [option]                 OPENSSL_NO_GOST
    idea                [option]                 OPENSSL_NO_IDEA (skip crypto\idea)
    ktls                [default]                OPENSSL_NO_KTLS
    legacy              [option]                 
    makedepend          [option]                 
    md2                 [default]                OPENSSL_NO_MD2 (skip crypto\md2)
    md4                 [option]                 OPENSSL_NO_MD4 (skip crypto\md4)
    mdc2                [option]                 OPENSSL_NO_MDC2 (skip crypto\mdc2)
    module              [cascade]                
    msan                [default]                OPENSSL_NO_MSAN
    padlockeng          [cascade]                OPENSSL_NO_PADLOCKENG
    poly1305            [option]                 OPENSSL_NO_POLY1305 (skip crypto\poly1305)
    rc2                 [option]                 OPENSSL_NO_RC2 (skip crypto\rc2)
    rc4                 [option]                 OPENSSL_NO_RC4 (skip crypto\rc4)
    rc5                 [option]                 OPENSSL_NO_RC5 (skip crypto\rc5)
    rmd160              [option]                 OPENSSL_NO_RMD160 (skip crypto\ripemd)
    scrypt              [option]                 OPENSSL_NO_SCRYPT
    sctp                [default]                OPENSSL_NO_SCTP
    seed                [option]                 OPENSSL_NO_SEED (skip crypto\seed)
    siphash             [option]                 OPENSSL_NO_SIPHASH (skip crypto\siphash)
    sm2                 [option]                 OPENSSL_NO_SM2 (skip crypto\sm2)
    sm3                 [option]                 OPENSSL_NO_SM3 (skip crypto\sm3)
    sm4                 [option]                 OPENSSL_NO_SM4 (skip crypto\sm4)
    srp                 [option]                 OPENSSL_NO_SRP (skip crypto\srp)
    srtp                [option]                 OPENSSL_NO_SRTP
    ssl-trace           [default]                OPENSSL_NO_SSL_TRACE
    tests               [option]                 OPENSSL_NO_TESTS
    trace               [default]                OPENSSL_NO_TRACE
    ts                  [option]                 OPENSSL_NO_TS (skip crypto\ts)
    ubsan               [default]                OPENSSL_NO_UBSAN
    ui-console          [option]                 OPENSSL_NO_UI_CONSOLE
    unit-test           [default]                OPENSSL_NO_UNIT_TEST
    whirlpool           [option]                 OPENSSL_NO_WHIRLPOOL (skip crypto\whrlpool)
    weak-ssl-ciphers    [default]                OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]                
    zlib-dynamic        [default]                
    ssl3                [option(ssl3-method)]    OPENSSL_NO_SSL3
    ssl3-method         [option]                 OPENSSL_NO_SSL3_METHOD
    tls1                [option(tls1-method)]    OPENSSL_NO_TLS1
    tls1-method         [option]                 OPENSSL_NO_TLS1_METHOD
    tls1_1              [option(tls1_1-method)]  OPENSSL_NO_TLS1_1
    tls1_1-method       [option]                 OPENSSL_NO_TLS1_1_METHOD
    tls1_3              [option]                 OPENSSL_NO_TLS1_3
    dtls1               [option(dtls1-method)]   OPENSSL_NO_DTLS1
    dtls1-method        [option]                 OPENSSL_NO_DTLS1_METHOD
    dtls1_2             [option(dtls1_2-method)] OPENSSL_NO_DTLS1_2
    dtls1_2-method      [option]                 OPENSSL_NO_DTLS1_2_METHOD

Config target attributes:

    AR => "lib",
    ARFLAGS => "/nologo",
    AS => "ml",
    ASFLAGS => "/nologo /Zi",
    CC => "cl",
    CFLAGS => "/W3 /wd4090 /nologo /O2",
    CPP => "\$(CC) /EP /C",
    HASHBANGPERL => "/usr/bin/env perl",
    LD => "link",
    LDFLAGS => "/nologo /debug",
    MT => "mt",
    MTFLAGS => "-nologo",
    RANLIB => "CODE(0xe42140)",
    RC => "rc",
    aroutflag => "/out:",
    asflags => "/Cp /coff /c /Cx",
    asm_arch => "x86",
    asoutflag => "/Fo",
    bin_cflags => "/Zi /Fdapp.pdb",
    bin_lflags => "/subsystem:console /opt:ref",
    bn_ops => "BN_LLONG",
    build_file => "makefile",
    build_scheme => [ "unified", "windows", "VC-common", "VC-WOW" ],
    cflags => "/Gs0 /GF /Gy /MD",
    coutflag => "/Fo",
    cppflags => "",
    defines => [ "OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN", "UNICODE", "_UNICODE", "_CRT_SECURE_NO_DEPRECATE", "_WINSOCK_DEPRECATED_NO_WARNINGS" ],
    disable => [  ],
    dso_cflags => "/Zi /Fddso.pdb",
    dso_scheme => "win32",
    enable => [  ],
    ex_libs => "ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib",
    includes => [  ],
    ldoutflag => "/out:",
    lflags => "",
    lib_cflags => "/Zi /Fdossl_static.pdb",
    lib_cppflags => "",
    lib_defines => [ "L_ENDIAN" ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "/dll",
    mtinflag => "-manifest ",
    mtoutflag => "-outputresource:",
    perl_platform => "Windows::MSVC",
    perlasm_scheme => "win32",
    rcoutflag => "/fo",
    shared_cflag => "",
    shared_defflag => "",
    shared_defines => [  ],
    shared_ldflag => "/dll",
    shared_rcflag => "",
    shared_target => "win-shared",
    sys_id => "WIN32",
    thread_defines => [  ],
    thread_scheme => "winthreads",
    unistd => "<unistd.h>",
    uplink_arch => "common",

Recorded environment:

    AR = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPPFLAGS = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LDFLAGS = 
    LDLIBS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = lib
    ARFLAGS         = /nologo
    AS              = ml
    ASFLAGS         = /nologo /Zi
    CC              = cl
    CFLAGS          = /W3 /wd4090 /nologo /O2 -Os
    CPP             = $(CC) /EP /C
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXXFLAGS        = -Os
    HASHBANGPERL    = /usr/bin/env perl
    LD              = link
    LDFLAGS         = /nologo /debug
    LDLIBS          = 
    MT              = mt
    MTFLAGS         = -nologo
    PERL            = C:\Perl64\bin\perl.exe
    RC              = rc
    RCFLAGS         = 

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    makefile

build file:

    makefile

build file templates:

    Configurations\common0.tmpl
    Configurations\windows-makefile.tmpl
    Configurations\common.tmpl
```

**Comment by misery** [comment]
Looks like it tries to install ```modules``` but feature ```module``` is disabled.

**Comment by mattcaswell** [comment]
Does it work if you don't use `no-dso`?

Ping @levitte.

**Comment by levitte** [comment]
> Looks like it tries to install ```modules``` but feature ```module``` is disabled.

That in itself shouldn't matter.

**Comment by levitte** [comment]
@misery, could you find the lines starting with `INSTALL_MODULES=` and  `INSTALL_ENGINES=` in `makefile` and display it here?

**Comment by levitte** [comment]
Actually, never mind, I found it on my own...

```
INSTALL_MODULES=Program fragment delivered error ``Undefined subroutine &OpenSSL::safe::fill_lines called at ..\openssl\Configurations\windows-makefile.tmpl line 103.''
```

Will fix

**Comment by levitte** [comment]
Fix in #11629, please try it out

**Comment by misery** [comment]
Thanks, we applied the patch and it works!
